src/ast.rs

```diff
@@ -0,0 +1,364 @@
+#[derive(Debug, Eq, PartialEq)]
+pub struct Assign {
+    pub variable: Variable,
+    pub value: Expression,
+}
+
+impl From<Assign> for Statement {
+    fn from(statement: Assign) -> Statement {
+        Statement::Assign(statement)
+    }
+}
+
+/// The name and value of an attribute
+#[derive(Debug, Eq, PartialEq)]
+pub struct Attribute {
+    pub name: Identifier,
+    pub value: Expression,
+}
+
+/// An `edge` statement that creates a new edge
+#[derive(Debug, Eq, PartialEq)]
+pub struct CreateEdge {
+    pub source: Expression,
+    pub sink: Expression,
+}
+
+impl From<CreateEdge> for Statement {
+    fn from(statement: CreateEdge) -> Statement {
+        Statement::CreateEdge(statement)
+    }
+}
+
+/// A `node` statement that creates a new graph node
+#[derive(Debug, Eq, PartialEq)]
+pub struct CreateGraphNode {
+    pub node: Variable,
+}
+
+impl From<CreateGraphNode> for Statement {
+    fn from(statement: CreateGraphNode) -> Statement {
+        Statement::CreateGraphNode(statement)
+    }
+}
+
+/// A `let` statement that declares a new immutable variable
+#[derive(Debug, Eq, PartialEq)]
+pub struct DeclareImmutable {
+    pub variable: Variable,
+    pub value: Expression,
+}
+
+impl From<DeclareImmutable> for Statement {
+    fn from(statement: DeclareImmutable) -> Statement {
+        Statement::DeclareImmutable(statement)
+    }
+}
+
+/// A `var` statement that declares a new mutable variable
+#[derive(Debug, Eq, PartialEq)]
+pub struct DeclareMutable {
+    pub variable: Variable,
+    pub value: Expression,
+}
+
+impl From<DeclareMutable> for Statement {
+    fn from(statement: DeclareMutable) -> Statement {
+        Statement::DeclareMutable(statement)
+    }
+}
+
+/// A `print` statement that prints out some debugging information
+#[derive(Debug, Eq, PartialEq)]
+pub struct Print {
+    pub values: Vec<Expression>,
+}
+
+impl From<Print> for Statement {
+    fn from(statement: Print) -> Statement {
+        Statement::Print(statement)
+    }
+}
+
+/// A `scan` statement that matches regular expressions against a string
+#[derive(Debug, Eq, PartialEq)]
+pub struct Scan {
+    pub value: Expression,
+    pub arms: Vec<ScanArm>,
+}
+
+impl From<Scan> for Statement {
+    fn from(statement: Scan) -> Statement {
+        Statement::Scan(statement)
+    }
+}
+
+/// One arm of a `scan` statement
+#[derive(Debug)]
+pub struct ScanArm {
+    pub regex: Regex,
+    pub statements: Vec<Statement>,
+}
+
+impl Eq for ScanArm {}
+
+impl PartialEq for ScanArm {
+    fn eq(&self, other: &ScanArm) -> bool {
+        self.regex.as_str() == other.regex.as_str() && self.statements == other.statements
+    }
+}
+
+/// A reference to a variable
+#[derive(Debug, Eq, PartialEq)]
+pub enum Variable {
+    Scoped(ScopedVariable),
+    Unscoped(UnscopedVariable),
+}
+
+/// A reference to a scoped variable
+#[derive(Debug, Eq, PartialEq)]
+pub struct ScopedVariable {
+    pub scope: Box<Expression>,
```

Comment:
Do we need arbitrary expressions as the scope of an identifier? e.g. string/integer scopes seem surprising?

Reply:
It has to be an expression that evaluates to a syntax node reference, but it's checked at runtime: https://github.com/tree-sitter/tree-sitter-graph/blob/3550c82d382ebd329dc54eee10dcfcef29439c2e/src/execution.rs#L538-L541